### ISSUE TITLE
[EPIC] Search 

### ISSUE THREAD

**Issue by juan-guerschman** [TerriaJS/terriajs#3286]
apologies if this is already open in some other ticket, if so I couldn't find it.

I want to quickly find the SA2: 201031017  in https://nationalmap.gov.au/#share=s-w89XMVg2poL5Hvm2HUU8IVwoTdV 

There are too many so clicking one by one until finding it is not practical. 
It would be good if the user could open a table and list all elements in the layer and select them in such a table. 

Is such a thing possible?

**Comment by AnaBelgun** [comment]
Hi @juan-guerschman 
to understand the issue better
1. do you mean downloading the Feature Info panel - currently that's possible as csv/json
2. do you mean once you added the boundaries, to search by SA2 numbers, rather than name (which would take you to the location/area)
3. or do you mean something completely different e.g. making possible to download any data (inlcuding WMS, WFS etc) as a CSV/tabular 
we had some discussions about similar issues here: https://github.com/TerriaJS/nationalmap/issues/174
let me know

**Comment by juan-guerschman** [comment]
I mean #2 
being able to find a SA2 element by its name or any other of the elements
![image](https://user-images.githubusercontent.com/8801702/56018329-27995e80-5d45-11e9-89d3-2872b390918d.png)

**Comment by AnaBelgun** [comment]
@kring is this a more generic "search within a WMS/WFS" feature?

**Comment by kring** [comment]
> is this a more generic "search within a WMS/WFS" feature?

Yes, I think that's what it is.

I don't think this is a "this sprint" thing, but I'll assign it to Ana so she can decide.

**Comment by AnaBelgun** [comment]
yep, definitely not this sprint (with 30+ issues on the list). I'll leave it in the product backlog for now.

**Comment by rowanwins** [comment]
Dropping a few thoughts on this @AnaBelgun 

- It would be worthing thinking about whether we just wanted to provision search on just some of the typical admin boundaries layers (eg for SA1, SA2 etc) vs whether we want people to search any data that they add to the map
  - being able to search on a few limited things would be simpler than working out how to support search on any layer that gets added

- If we were going for the later (to support any layer which a user added) we'd need to implement different approaches for different layer types (eg you'd query WMS vs WFS differently) so we'd want to pick some priority data source types.

- The other challenge with the approach of supporting searching on any layer is knowing what attribute/property of the data someone wants to search on as we'd want to avoid having to do a global search on the data... so if we went down this route we'd probably need to introduce some new UI pieces that don't yet exist.
  - if we went for just a basic search on admin boundaries we might be able to get away with dropping the results underneath the existing search results

Does that provide some food for thought?

**Comment by meh9** [comment]
Just linking in some previous related issues, also see issues linked from those: 
#1324 
#1196

**Comment by AnaBelgun** [comment]
I think he generic option is better (than implement search for boundaries) and from what we have heard from users, they would love to be able to search on CSVs (i.e. elements in the table) (as 1st priority, or more popular than other types of data) and then WMS and WFS (as second in popularity). If we support search within datasets, the search UI will have to change for sure., even if minimal. 

To recap: 
- we should support general search within datasets
- start with CSVs and determine what elements and prioritise
- move to WMS and WFS 
- change the UI to reflect search within data

Let's see what @meh9 and @kring think about it; will then have to assess the complexity and how long will take to finalise prioritisation

**Comment by kring** [comment]
It's not a small feature, but it sounds reasonable enough to me. Code-wise, we could add a new search provider that gives every catalog item on the workbench the opportunity to provide search results via a new (and optional) method. Then we can just incrementally implement that method for different types of catalog items over time. Catalog items already have a mechanism for determining the "name" property for feature info purposes; we could use similar logic to enable searching by name. If we want more sophisticated searching, though, it starts to sound more like the vector data filtering feature and could get a lot more complicated.

**Comment by AnaBelgun** [comment]
We decided to break this into: 

- [ ] search within WMS and WFS - #3385 

- [ ] search within CSV - https://github.com/TerriaJS/terriajs/issues/3386

- [ ] search within other (TBC)

**Comment by philipgrimmett** [comment]
Invision concept:
https://invis.io/GHRZ22948RC#/362867174_workbench_With_Data

Assumptions:
- Data results should appear before locations
- Active data results should appear before catalogue results

Now might be a good time to discuss if we should continue to display Locations, Official Place Names, and Addresses separately. There seem to be a lot of (what look like) duplicate results that take up a lot of screen real estate, which means we'll probably have to collapse data results by default so locations aren't pushed below the fold.

**Comment by philipgrimmett** [comment]
**Feedback via DTA**
> I am a frequent user of the Natgitional Map website (https://nationalmap.gov.au/)
> I was wondering if there is the ability to search (or zoom into) a known sa2 number (eg 309061248) without knowing the sa2 name or address location.
>Any assistance would be greatly appreciated.

**Comment by AnaBelgun** [comment]
Closed in favour of https://github.com/TerriaJS/product/issues/555. A new issue with OSS requirements to be submitted.
